sdk/core/Azure.Core/tests/TestFramework/RecordMatcher.cs

```diff
@@ -41,39 +43,98 @@ public virtual int FindMatch(Request request, IList<RecordEntry> entries, out st
             string uri = _sanitizer.SanitizeUri(request.UriBuilder.ToString());
 
             int bestScore = int.MaxValue;
-            var messagePrefix = failureMessage = $"Unable to find recorded request for {request.Method} {request.UriBuilder.ToString()}";
+            RecordEntry bestScoreEntry = null;
 
-            for (int i = 0; i < entries.Count; i++)
+            foreach (RecordEntry entry in entries)
             {
-                RecordEntry entry = entries[i];
+                int score = 0;
+                if (entry.RequestUri != uri)
+                {
+                    score++;
+                }
 
-                if (entry.RequestUri == uri && entry.RequestMethod == request.Method)
+                if (entry.RequestMethod != request.Method)
                 {
-                    int score = CompareHeaderDictionaries(headers, entry.RequestHeaders, out var diff);
-                    if (score == 0)
-                    {
-                        failureMessage = null;
-                        return i;
-                    }
-                    else if (score < bestScore)
+                    score++;
+                }
+
+                score += CompareHeaderDictionaries(headers, entry.RequestHeaders);
+
+                if (score == 0)
+                {
+                    return entry;
+                }
+
+                if (score < bestScore)
+                {
+                    bestScoreEntry = entry;
+                    bestScore = score;
+                }
+            }
+
+
+            throw new InvalidOperationException(GenerateException(request.Method, uri, headers, bestScoreEntry));
+        }
+
+        private string GenerateException(RequestMethod requestMethod, string uri, SortedDictionary<string, string[]> headers, RecordEntry bestScoreEntry)
+        {
+            StringBuilder builder = new StringBuilder();
+            builder.AppendLine($"Unable to find a record for the request {requestMethod} {uri}");
+
+            if (bestScoreEntry == null)
+            {
+                builder.AppendLine("No records to match.");
+                return builder.ToString();
+            }
+
+            if (requestMethod != bestScoreEntry.RequestMethod)
+            {
+                builder.AppendLine($"Method doesn't match, request <{requestMethod}> record <{bestScoreEntry.RequestMethod}>");
+            }
+
+            if (uri != bestScoreEntry.RequestUri)
+            {
+                builder.AppendLine($"Uri doesn't match, request <{uri}>, record <{bestScoreEntry.RequestUri}>");
```

Comment:
Nit - could we do this as two lines with the URIs aligned?  I've had to paste a lot of these into notepad and align them by hand to see where things went wrong.